<a href="https://www.kaggle.com/code/erkanhatipoglu/heart-failure-prediction-with-xgboost?scriptVersionId=106643658" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Introduction  <a id='introduction'></a>

This is a starter code for Heart Failure Prediction using XGBoost. By forking the notebook, the reader may quickly try different approaches and concepts (EDA, Feature Engineering, other models, etc.). I have tried to keep this notebook as basic as possible.

Sklearn has several functions to calculate and display accuracy, recall, precision, and F1 scores. However, I will not use those functions and will calculate those results manually for demonstration purposes. 

We need to increase the recall score for this dataset task since we do not want to misclassify heart failure cases. In other words, we want the number of true positives to be as high as possible while the number of false negatives is as small as possible. It is better to falsely treat a patient than to leave another patient to die because of misclassification.

It seems that the time column needs to be dropped since it is not a predictor. It turns out that we cannot get a time value for new patients after deployment. Please refer to [this thread](https://www.kaggle.com/andrewmvd/heart-failure-clinical-data/discussion/178372) and [Attribute Information](https://archive.ics.uci.edu/ml/datasets/Heart+failure+clinical+records) for further discussions on the subject.

I also want to thank [@Larxel](https://www.kaggle.com/andrewmvd) for this dataset.

Thank you for reading.


# Table of Contents
* [Introduction](#introduction)
* [Helper Functions](#functions)
* [Getting the Data](#getdata)
* [Preprocessing](#preprocessing) 
* [Pipeline and Validation with early_stopping_rounds](#pipeline) 
* [Grid Search](#gridsearch)
* [Cross-validation](#cross-validation)
* [Prediction](#prediction)
* [Conclusion](#conclusion) 
* [References](#references)

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

from pandas_profiling import ProfileReport
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, classification_report, recall_score
import pandas_profiling
from sklearn.feature_selection import SelectKBest

/kaggle/input/heart-failure-clinical-data/heart_failure_clinical_records_dataset.csv


# Helper Functions <a id='functions'></a>  

<div class="alert alert-block alert-info">
<b>Tip:</b> We will use some helper functions throughout the notebook. Collecting them in one place is a good idea, making the code more organized.
</div>

In [2]:
def calculate_cf (y_valid, preds):
    ''' A function that calculates the confusion matrix, accuracy, precision, recall,
    and f1_score. Accuracy, precision, recall, and f1_score can be easily obtained by
    using sklearn features.This function is only for demontration purposes.'''
    
    # Calculating confusion_matrix
    CM = confusion_matrix(y_valid, preds)

    # Calculate True Positives(TP), False Positives(FP)
    # False Negative(FN) and True Negatives(TN) from confusion_matrix
    true_negatives = CM[0][0]
    false_negatives = CM[1][0]
    true_positives = CM[1][1]
    false_positives = CM[0][1]

    # You can easily get these values in sklearn using 
    # accuracy_score, precision_score, classification_report, etc.
    # I calculate these values for demonstration purposes.
    accuracy = (true_positives + true_negatives)/(true_positives + false_positives + false_negatives + true_negatives)
    precision = (true_positives) / (true_positives + false_positives)
    recall = (true_positives) / (true_positives + false_negatives)
    f1_score = 2 * (precision * recall) / (precision + recall)
    
    return true_negatives, false_negatives, true_positives, false_positives,accuracy, precision, recall, f1_score

def display_results (y_valid, preds):
    ''' A function that displays the results'''
    # get the results
    true_negatives, false_negatives, true_positives, false_positives,accuracy, precision, recall, f1_score = calculate_cf (y_valid, preds)
    
    blank= " "
    star = "*"
    print(blank*50 + "Death Event") 
    print(blank*30 + star*55)
    print(blank*35 + "Positive" + blank * 30 + "Negative")
    print(blank * 30 + star*18 + blank * 19 + star*18)
    print("Predicted" + blank*7 + "Positive" + blank  * 11 + str(true_positives)+ " (TP)" + blank * 32 + str(false_positives) + " (FP)")
    print("Class" + blank*11 + "Negative" + blank  * 11 + str(false_negatives)+ " (FN)" + blank * 31 + str(true_negatives) + " (TN)")
    print()
    print("Accuracy = (TP+TN)/(TP+FP+FN+TN) = {:.2f}".format(accuracy))
    print("Precision = (TP)/(TP+FP) = {:.2f}".format(precision))
    print("Recall = (TP)/(TP+FN) = {:.2f}".format(recall))
    print("F1 score = 2 * (Precision*Recall)/(Precision+Recall) = {:.2f}".format(f1_score))

# Getting the Data  <a id='getdata'></a>

In [3]:
# Read the data
train_data = pd.read_csv('/kaggle/input/heart-failure-clinical-data/heart_failure_clinical_records_dataset.csv')

# Make a copy to avoid changing original data
X = train_data.copy()

print('Data is OK')

Data is OK


In [4]:
X.head()

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

In [5]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       299 non-null    float64
 1   anaemia                   299 non-null    int64  
 2   creatinine_phosphokinase  299 non-null    int64  
 3   diabetes                  299 non-null    int64  
 4   ejection_fraction         299 non-null    int64  
 5   high_blood_pressure       299 non-null    int64  
 6   platelets                 299 non-null    float64
 7   serum_creatinine          299 non-null    float64
 8   serum_sodium              299 non-null    int64  
 9   sex                       299 non-null    int64  
 10  smoking                   299 non-null    int64  
 11  time                      299 non-null    int64  
 12  DEATH_EVENT               299 non-null    int64  
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


In [6]:
X.describe()

age     anaemia  creatinine_phosphokinase    diabetes  \
count  299.000000  299.000000                299.000000  299.000000   
mean    60.833893    0.431438                581.839465    0.418060   
std     11.894809    0.496107                970.287881    0.494067   
min     40.000000    0.000000                 23.000000    0.000000   
25%     51.000000    0.000000                116.500000    0.000000   
50%     60.000000    0.000000                250.000000    0.000000   
75%     70.000000    1.000000                582.000000    1.000000   
max     95.000000    1.000000               7861.000000    1.000000   

       ejection_fraction  high_blood_pressure      platelets  \
count         299.000000           299.000000     299.000000   
mean           38.083612             0.351171  263358.029264   
std            11.834841             0.478136   97804.236869   
min            14.000000             0.000000   25100.000000   
25%            30.000000             0.000000  212500.000000   
50%            38.000000             0.000000  262000.000000   
75%            45.000000             1.000000  303500.000000   
max            80.000000             1.000000  850000.000000   

       serum_creatinine  serum_sodium         sex    smoking        time  \
count         299.00000    299.000000  299.000000  299.00000  299.000000   
mean            1.39388    136.625418    0.648829    0.32107  130.260870   
std             1.03451      4.412477    0.478136    0.46767   77.614208   
min             0.50000    113.000000    0.000000    0.00000    4.000000   
25%             0.90000    134.000000    0.000000    0.00000   73.000000   
50%             1.10000    137.000000    1.000000    0.00000  115.000000   
75%             1.40000    140.000000    1.000000    1.00000  203.000000   
max             9.40000    148.000000    1.000000    1.00000  285.000000   

       DEATH_EVENT  
count    299.00000  
mean       0.32107  
std        0.46767  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        1.00000  
max        1.00000

# Preprocessing  <a id='preprocessing'></a>

In [7]:
# get the target, separate target and time from predictors
y = X.DEATH_EVENT              
X.drop(['DEATH_EVENT', 'time'], axis=1, inplace=True)

# Break off validation set from training data
X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size=0.8, test_size=0.2,
                                                                random_state=0)
# Display results
print ("Shapes:")
print ("X: {}".format(X.shape))
print ("y: {}".format(y.shape))
print()
print ("X_train: {}".format(X_train.shape))
print ("X_valid: {}".format(X_valid.shape))
print ("y_train: {}".format(y_train.shape))
print ("y_valid: {}\n".format(y_valid.shape))

Shapes:
X: (299, 11)
y: (299,)

X_train: (239, 11)
X_valid: (60, 11)
y_train: (239,)
y_valid: (60,)



In [8]:
# Select numeric columns
numerical_cols = [cname for cname in X_train.columns if X_train[cname].dtype in ['int64', 'float64']]
numerical_cols

['age',
 'anaemia',
 'creatinine_phosphokinase',
 'diabetes',
 'ejection_fraction',
 'high_blood_pressure',
 'platelets',
 'serum_creatinine',
 'serum_sodium',
 'sex',
 'smoking']

# Pipeline and Validation with early_stopping_rounds <a id='pipeline'></a>

In [9]:
# Define transformers
# Preprocessing for numerical data

numerical_transformer = Pipeline(steps=[('scaler', StandardScaler())])

# Bundle preprocessing for numerical and categorical data
col_trans = ColumnTransformer(
    transformers=[('numerical', numerical_transformer, numerical_cols)])

select = SelectKBest(k=11)

preprocessor= Pipeline(steps=[('column_tarnsformer', col_trans), ('selection', select)])

In [10]:
# Define Model
model = XGBClassifier(learning_rate = 0.08,
                            n_estimators=500,
                            max_depth=5,
                            min_child_weight=1,
                            gamma=0,
                            subsample=0.8,
                            colsample_bytree=0.8,
                            reg_alpha = 0,
                            reg_lambda = 1,
                            random_state=42)

<div class="alert alert-block alert-danger">  
<p>If we want to use early_stopping_rounds with our pipeline, we cannot use the validation set (X_valid) directly. This is because sklearn pipelines do not process the eval_set used with early_stopping_rounds. As a result, we need to process our validation set before using early_stopping_rounds.</p> 

<p>There is a great danger here. If we forget to process our validation set and if processed data has the same number of columns as the unprocessed data, we may not see an error. Validation with unprocessed data may mislead us.</p>

<p>To process the eval_set, we need to fit_transform X_valid by using our preprocessor, a pipeline consisting of transformers that do not have an estimator.</p>
</div>

In [11]:
# Preprocessing of validation data
X_valid_eval = preprocessor.fit(X_train, y_train).transform (X_valid)

In [12]:
# Display the number of remaining columns after transformation 
print("We have", X_valid_eval.shape[1], "features after transformation")

We have 11 features after transformation


In [13]:
# Define XGBClassifier fitting parameters for the pipeline
fit_params = {"model__early_stopping_rounds": 50,
              "model__eval_set": [(X_valid_eval, y_valid)],
              "model__verbose": True,
              "model__eval_metric" : "error"}

In [14]:
# Create and Evaluate the Pipeline
# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', model)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train, **fit_params)

# Get predictions
preds = my_pipeline.predict(X_valid)

# Evaluate the model
score = accuracy_score(y_valid,preds)

# Get best estimator count
# https://stackoverflow.com/questions/51955256/xgboost-best-iteration
# thanks to @Alessandro Ceccarelli, @Eran Moshe
rounds = model.get_booster().best_ntree_limit

# Display the result
print("Score: {}".format(score))
print("Best number of trees: {}".format(rounds))


[0]	validation_0-error:0.26667
Will train until validation_0-error hasn't improved in 50 rounds.
[1]	validation_0-error:0.26667
[2]	validation_0-error:0.28333
[3]	validation_0-error:0.31667
[4]	validation_0-error:0.28333
[5]	validation_0-error:0.31667
[6]	validation_0-error:0.28333
[7]	validation_0-error:0.26667
[8]	validation_0-error:0.30000
[9]	validation_0-error:0.28333
[10]	validation_0-error:0.26667
[11]	validation_0-error:0.26667
[12]	validation_0-error:0.30000
[13]	validation_0-error:0.31667
[14]	validation_0-error:0.31667
[15]	validation_0-error:0.31667
[16]	validation_0-error:0.30000
[17]	validation_0-error:0.25000
[18]	validation_0-error:0.28333
[19]	validation_0-error:0.31667
[20]	validation_0-error:0.31667
[21]	validation_0-error:0.30000
[22]	validation_0-error:0.26667
[23]	validation_0-error:0.30000
[24]	validation_0-error:0.26667
[25]	validation_0-error:0.31667
[26]	validation_0-error:0.30000
[27]	validation_0-error:0.28333
[28]	validation_0-error:0.26667
[29]	validation_

In [15]:
# Display results
display_results (y_valid, preds)

                                                  Death Event
                              *******************************************************
                                   Positive                              Negative
                              ******************                   ******************
Predicted       Positive           11 (TP)                                3 (FP)
Class           Negative           12 (FN)                               34 (TN)

Accuracy = (TP+TN)/(TP+FP+FN+TN) = 0.75
Precision = (TP)/(TP+FP) = 0.79
Recall = (TP)/(TP+FN) = 0.48
F1 score = 2 * (Precision*Recall)/(Precision+Recall) = 0.59


# Grid Search  <a id='gridsearch'></a>

In [16]:
# Define model parameters for grid search
param_grid = {'model__learning_rate': [0.01, 0.05, 0.08, 0.1],
              'model__n_estimators': [rounds],
              'model__max_depth': [3, 4, 5, 6],
              'model__min_child_weight': [1, 2, 3, 4],
              'model__gamma': [0],
              'model__subsample': [0.60, 0.70, 0.80, 0.90],
              'model__colsample_bytree': [0.60, 0.70, 0.80, 0.90],
              'model__random_state' : [42]}

In [17]:
# Perform grid search
# Use model parameters defined above. We use scoring as recall to minimize 
# False Negatives
search = GridSearchCV(my_pipeline, param_grid, cv=5, n_jobs=-1,scoring='recall')
search.fit(X, y)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        Pipeline(steps=[('column_tarnsformer',
                                                         ColumnTransformer(transformers=[('numerical',
                                                                                          Pipeline(steps=[('scaler',
                                                                                                           StandardScaler())]),
                                                                                          ['age',
                                                                                           'anaemia',
                                                                                           'creatinine_phosphokinase',
                                                                                           'diabetes',
                                                                            

# Cross-validation   <a id='cross-validation'></a>

In [18]:
# Preprocessing of training data, fit model 
X_cv = X.copy()
y_cv = y.copy()

# X_cv = X_train.copy()
# y_cv = y_train.copy()

# get scores
scores = cross_val_score(search, X_cv, y_cv,
                              cv=5,
                              scoring='accuracy')

print("Accuracy:\n", scores)
print("Accuracy mean: {}".format(scores.mean()))
print("Accuracy std: {}".format(scores.std()))

Accuracy:
 [0.66666667 0.71666667 0.73333333 0.78333333 0.72881356]
Accuracy mean: 0.7257627118644068
Accuracy std: 0.03729900543501551


# Prediction  <a id='prediction'></a>

In [19]:
# Get predictions
preds = search.predict(X_valid)

# Evaluate the model
score = accuracy_score(y_valid,preds)

print("Score: {}".format(score))

Score: 0.8333333333333334


In [20]:
display_results (y_valid, preds)

                                                  Death Event
                              *******************************************************
                                   Positive                              Negative
                              ******************                   ******************
Predicted       Positive           16 (TP)                                3 (FP)
Class           Negative           7 (FN)                               34 (TN)

Accuracy = (TP+TN)/(TP+FP+FN+TN) = 0.83
Precision = (TP)/(TP+FP) = 0.84
Recall = (TP)/(TP+FN) = 0.70
F1 score = 2 * (Precision*Recall)/(Precision+Recall) = 0.76


# Conclusion  <a id='conclusion'></a>

<div class="alert alert-block alert-warning">  
<b>Warning:</b> Below, for the XGBClassifier(random_state=0) example, we need to change both L10 and L16 for the model and L7 for the train_test_split.  
</div>

We had used `scoring='recall'` for our grGridSearchCV, and improved our scores as below:

### Using train_test_split(random_state=0)

* Accuracy:  from 0.77 to 0.85
* Precision: from 0.76 to 0.94
* Recall:    from 0.57 to 0.65
* F1 score:  from 0.65 to 0.77

If we had shuffled our validation set with different random_state values we would have got different results such as:


### Using train_test_split(random_state=1)
*(change L7)*

* Accuracy:  from 0.85 to 0.92
* Precision: from 0.73 to 0.85
* Recall:    from 0.57 to 0.79
* F1 score:  from 0.64 to 0.81

### Using train_test_split(random_state=42)
*(change L7)*

* Accuracy:  from 0.72 to 0.80
* Precision: from 0.72 to 0.84
* Recall:    from 0.52 to 0.64
* F1 score:  from 0.60 to 0.73


If we had changed our model with different random_state values we would have got different results such as:

### Using XGBClassifier(random_state=0)
*(change L7, L10, L16)*

* Accuracy:  from 0.75 to 0.92
* Precision: from 0.79 to 0.95
* Recall:    from 0.48 to 0.83
* F1 score:  from 0.59 to 0.88

# References   <a id='references'></a>
* [@Larxel](https://www.kaggle.com/andrewmvd)
* [Do not use time as a feature (Discussion Thread)](https://www.kaggle.com/andrewmvd/heart-failure-clinical-data/discussion/178372)
* [Machine learning can predict survival of patients with heart failure from serum creatinine and ejection fraction alone (Research Article)](https://bmcmedinformdecismak.biomedcentral.com/articles/10.1186/s12911-020-1023-5)
* [10-simple-hacks-to-speed-up-your-data-analysis - Parul Pandey](https://www.kaggle.com/parulpandey/10-simple-hacks-to-speed-up-your-data-analysis)
* [Dataset Transformations - Scikit-learn](https://scikit-learn.org/stable/data_transforms.html)
* [Intermediate Machine Learning Course - Pipelines](https://www.kaggle.com/alexisbcook/pipelines)
* [Kaggle Learn](https://www.kaggle.com/learn/overview)